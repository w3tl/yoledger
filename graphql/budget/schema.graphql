extend type Query {
  budgets(dateStart: Date!, count: Int = 5): BudgetPlan
}
extend type Mutation {
  upsertBudget(input: UpsertBudgetInput!): UpsertBudgetPayload
}

type BudgetPlan {
  periods: [Date]
  accounts: [Account]
  budget: [Budget]
}

type Budget {
  date: Date!
  allocations: [Allocation]
}

type Allocation {
  account: Account!
  amount: Float
  balance: Float
}

input UpsertBudgetInput {
  account: String!
  date: Date!
  amount: Float!
}

type UpsertBudgetPayload {
  success: Boolean
}
