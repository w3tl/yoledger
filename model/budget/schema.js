export default {
  type: 'object',
  required: ['date', 'userId'],
  additionalProperties: false,
  properties: {
    _id: {
      bsonType: 'objectId',
    },
    date: {
      bsonType: 'date',
    },
    account: {
      type: 'string',
    },
    amount: {
      type: 'number',
    },
    balance: {
      type: 'number',
    },
    userId: {
      type: 'string',
    },
  },
};
