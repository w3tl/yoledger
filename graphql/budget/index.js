import Schema from './schema.graphql';
import {
  Query, Mutation, BudgetPlan,
} from './resolver';

export {
  Schema, Query, Mutation, BudgetPlan,
};
