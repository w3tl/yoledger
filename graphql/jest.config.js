module.exports = {
  notify: true,
  // testEnvironment: './graphql-environment.js',
  transform: {
    '^.+\\.js?$': 'babel-jest',
  },
};
