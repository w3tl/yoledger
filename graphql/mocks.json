{
  "users": [{
    "_id": "admin"
  }, {
    "_id": "user1"
  }],
  "transactions": [
    {
      "_id": "1",
      "amount": 100.5,
      "createdAt": "2018-06-20T00:00:00.000Z",
      "destination": {
        "name": "Food"
      },
      "source": {
        "name": "Bank card"
      }
    },
    {
      "_id": "2",
      "amount": 13,
      "createdAt": "2018-06-21T00:00:00.000Z",
      "destination": {
        "name": "Food"
      },
      "source": {
        "name": "Bank card"
      }
    }
  ],
  "accounts": [
    {
      "_id": "1",
      "balance": 1500.5,
      "name": "Bank card",
      "type": "ASSET"
    },
    {
      "_id": "2",
      "balance": 143,
      "name": "Cash",
      "type": "ASSET"
    },
    {
      "_id": "3",
      "name": "Food",
      "type": "EXPENSE"
    },
    {
      "_id": "4",
      "name": "Train",
      "type": "EXPENSE"
    }
  ],
  "budgets": [
    {
      "_id": "1",
      "account": "Food",
      "date": "2018-05-21T00:00:00.000Z",
      "amount": 50
    },
    {
      "_id": "2",
      "account": "Food",
      "date": "2018-05-10T00:00:00.000Z",
      "amount": 55
    },
    {
      "_id": "3",
      "account": "Food",
      "date": "2018-06-21T00:00:00.000Z",
      "amount": 40
    },
    {
      "_id": "4",
      "account": "Food",
      "date": "2018-06-10T00:00:00.000Z",
      "amount": 60
    },
    {
      "_id": "5",
      "account": "Train",
      "date": "2018-06-10T00:00:00.000Z",
      "amount": 14
    }
  ]
}
