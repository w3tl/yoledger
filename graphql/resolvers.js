import {
  Query as UserQuery,
  User,
} from './user';
import {
  Query as AccountQuery,
  Mutation as AccountMutation,
  Account,
} from './account';
import {
  Query as TransactionQuery,
  Mutation as TransactionMutation,
  Transaction,
} from './transaction';
import {
  Query as BudgetQuery,
  Mutation as BudgetMutation,
  Budget,
  BudgetPlan,
} from './budget';
import dateResolver from './date';

export default {
  Query: Object.assign({}, UserQuery, AccountQuery, TransactionQuery, BudgetQuery),
  Mutation: Object.assign({}, AccountMutation, TransactionMutation, BudgetMutation),
  User,
  Account,
  Transaction,
  Budget,
  BudgetPlan,
  Date: dateResolver,
};
