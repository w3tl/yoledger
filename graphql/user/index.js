import Schema from './schema.graphql';
import { Query, User } from './resolver';

export {
  Schema, Query, User,
};
