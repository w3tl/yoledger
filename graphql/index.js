import { makeExecutableSchema } from 'graphql-tools';
import typeDefs from './schema';
import resolvers from './resolvers';
import dataloaders from './dataloaders';

export default makeExecutableSchema({
  typeDefs,
  resolvers,
});

export { dataloaders };
