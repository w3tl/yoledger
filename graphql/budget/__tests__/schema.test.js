jest.mock('mongodb');

import { graphql } from 'graphql';
import schema, { dataloaders } from '../../index';

describe('budget schema', () => {
  const { connection } = require('mongodb');
  it('must return budget between dates', async () => {
    const query = `
    query budget($dateStart: Date!, $dateEnd: Date!) {
      budget(dateStart: $dateStart, dateEnd: $dateEnd) {
         account { name }
         date
         amount
      }
    }
    `;
    const rootValue = {};
    const context = { dataloaders: dataloaders(connection), userId: 'admin', connection };
    const variables = {
      dateStart: '2018-05-10',
      dateEnd: '2018-05-20',
    };

    const result = await graphql(schema, query, rootValue, context, variables);
    expect(result.errors).toBeUndefined();
    expect(result.data.budget).toMatchSnapshot(`between ${variables.dateStart} and ${variables.dateEnd}`);
  });

  it('must return budget for account between dates', async () => {
    const query = `
    query budget($dateStart: Date!, $dateEnd: Date!, $account: String) {
      budget(dateStart: $dateStart, dateEnd: $dateEnd, account: $account) {
         account { name }
         date
         amount
      }
    }
    `;
    const rootValue = {};
    const context = { dataloaders: dataloaders(connection), userId: 'admin', connection };
    const variables = {
      dateStart: '2018-06-21',
      dateEnd: '2018-07-01',
      account: 'Food',
    };

    const result = await graphql(schema, query, rootValue, context, variables);
    expect(result.errors).toBeUndefined();
    expect(result.data.budget).toMatchSnapshot(`between ${variables.dateStart} and ${variables.dateEnd} for ${variables.account}`);
  });
});
