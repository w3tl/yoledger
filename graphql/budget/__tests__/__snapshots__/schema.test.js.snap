// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`budget schema must return plan with periods, accounts and allocations: 4 periods from 2018-05-10 1`] = `
Object {
  "budgets": Object {
    "accounts": Array [
      Object {
        "name": "Food",
      },
      Object {
        "name": "Train",
      },
    ],
    "budget": Array [
      Object {
        "allocations": Array [
          Object {
            "account": Object {
              "name": "Food",
            },
            "amount": 50,
            "balance": 0,
          },
          Object {
            "account": Object {
              "name": "Train",
            },
            "amount": 0,
            "balance": 0,
          },
        ],
        "date": "2018-05-10T00:00:00.000Z",
      },
      Object {
        "allocations": Array [
          Object {
            "account": Object {
              "name": "Food",
            },
            "amount": 55,
            "balance": 0,
          },
          Object {
            "account": Object {
              "name": "Train",
            },
            "amount": 0,
            "balance": 0,
          },
        ],
        "date": "2018-05-25T00:00:00.000Z",
      },
      Object {
        "allocations": Array [
          Object {
            "account": Object {
              "name": "Food",
            },
            "amount": 60,
            "balance": 0,
          },
          Object {
            "account": Object {
              "name": "Train",
            },
            "amount": 14,
            "balance": 0,
          },
        ],
        "date": "2018-06-10T00:00:00.000Z",
      },
      Object {
        "allocations": Array [
          Object {
            "account": Object {
              "name": "Food",
            },
            "amount": 0,
            "balance": 0,
          },
          Object {
            "account": Object {
              "name": "Train",
            },
            "amount": 0,
            "balance": 0,
          },
        ],
        "date": "2018-06-25T00:00:00.000Z",
      },
    ],
    "periods": Array [
      "2018-05-10T00:00:00.000Z",
      "2018-05-25T00:00:00.000Z",
      "2018-06-10T00:00:00.000Z",
      "2018-06-25T00:00:00.000Z",
    ],
  },
}
`;
